lib/charms/opensearch/v0/opensearch_distro.py

```diff
@@ -305,13 +309,79 @@ def write_file(self, path: str, data: str, override: bool = True):
         with open(path, mode="w") as f:
             f.write(data)
 
+    @property
+    def _plugin(self):
+        """Returns the executable path. Should be overloaded, e.g. by the snap command."""
+        return f"{self.paths.home}/bin/opensearch-plugin"
+
+    def add_plugin_without_restart(self, plugin: str, batch: bool = False) -> bool:
+        """Add a plugin to this node. Restart must be managed in separated."""
+        try:
+            args = "install"
+            if batch:
+                args += " --batch"
+            args += f" {plugin}"
+            self._run_cmd(self._plugin, args)
+        except OpenSearchCmdError as e:
+            if "already exists" in e.stderr:
+                return
+            raise OpenSearchPluginError(f"Failed to install plugin {plugin}")
+
+    def remove_plugin_without_restart(self, plugin):
+        """Remove a plugin without restarting the node."""
+        try:
+            args = f"remove {plugin}"
+            self._run_cmd(self._plugin, args)
+        except OpenSearchCmdError as e:
+            if "not found" in e.stderr:
+                logger.info("Plugin {plugin} not found, leaving remove method")
+                return
+            raise OpenSearchPluginError(f"Failed to remove plugin {plugin}")
+
+    def list_plugins(self):
+        """List plugins."""
+        try:
+            self._run_cmd(self._plugin, "list")
+        except OpenSearchCmdError:
+            raise OpenSearchPluginError("Failed to list plugins")
+
+    @property
+    def _keystore(self):
+        return f"{self.paths.home}/bin/opensearch-keystore"
+
+    def add_to_keystore(self, key: str, value: str, force: bool = False):
+        """Adds a given key to the "opensearch" keystore."""
+        if not value:
+            raise OpenSearchKeystoreError("Missing keystore value")
+        args = "add " if not force else "add --force "
+        args += f"{key}"
+        try:
+            # Add newline to the end of the key, if missing
+            v = value + ("" if value[-1] == "\n" else "\n")
+            self._run_cmd(self._keystore, args, input=v)
```

Comment:
Why are we doing this? we should have well formatted keys from the get-go, what we store here should match what is stored in the secrets or peer rel data

Reply:
@Mehdi-Bendriss this check goes to the input in the subprocess, which must have a "\n" to actually pass the value; whereas the content of ```value``` comes from the S3 relation, which may not have it.
We are using ```input``` so we do not create a temporary file with the keys in plain text while passing them to the keystore. It is more appropriate to be prompted by opensearch-keystore and pass it as an input.